Twisted callback

whirm opened the pull request:
Do not merge yet, I'm extending this PR to include the complete callback usage removal from dispersy.

whirm commented:
retest this please

NielsZeilemaker commented on bootstrap.py:
```diff
@@ -137,48 +141,22 @@ def resolve(self, func, timeout=60.0, blocking=False):
+            reactor.callLater(0, func, True)
         else:
+            # TODO(emilon): What is this used for? Can I remove it?
             self._thread_counter += 1
```
Was used to create a unique name for each callback resolve instance.

NielsZeilemaker commented on tests/test_callback.py:
```diff
@@ -1,12 +1,17 @@
 class TestCallback(DispersyTestFunc):
 
-    @call_on_mm_thread
+    #@call_on_mm_thread
```
These need to be on the mm thread because of the yield. Now they are bascially not running.

Commit:
Implement TwistedCallback

Commit:
Make the tracker use the twisted callback

Commit:
Twistify the bootstrap and allow it to be disabled

Commit:
Refactor bootstrap resolution stuff so it works with the new twisted
callback, still some stuff pending to do, but its working just fine.

Commit:
Port unit tests classes to the new twisted callback

Commit:
Fix the tests

Commit:
Community import cleanup

NielsZeilemaker commented on dispersy.py:
```diff
@@ -319,16 +320,14 @@ def retry_until_success():
-                # unable to resolve all... retry until successful
-                self._callback.register(retry_until_success, delay=300.0)
+            # TODO(emilon): Since we moved to twisted, this call will not block, but all this code should be refactored soon anyway
+            self._callback.register(bootstrap.resolve, kargs=dict(func=on_results, timeout=timeout))
```
It has to block, i'm not sure how the search community fast walker thingy will respond to not having any bootstrap candidates.

NielsZeilemaker commented on tests/test_dynamicsettings.py:
```diff
@@ -22,6 +23,7 @@ def test_default_resolution(self):
 
         # NODE creates a message (should allow, because the default policy is PublicResolution)
         message = self._mm.create_dynamic_resolution_text("Message #%d" % 10, policy.implement(), 10)
+        #self._mm.give_message(message, self._mm)
```
I guess these commented lines can be removed

NielsZeilemaker commented:
It's looking great, I have just one remark. Maybe remove the Callback object alltogether, and merge some of its methods to the twistedcallback class.
We're never going back are we :smile:

Commit:
Further twistify the bootstrap process, untested

whirm commented:
@NielsZeilemaker yeah, I planned to do that when the bootstraping was twistified and the tracker working (doing the later now)

Commit:
Skip looping resolution if bootstrap is disabled too

Commit:
Make callback.call() thread safe.
This means TwistedCallback's contructor can't use it anymore to obtain
the reactor's thread ID hence from now on we need to make sure it always gets
instanced from there.

NielsZeilemaker commented on dispersy.py:
```diff
@@ -301,36 +297,23 @@ def on_results(success):
+            now = False
 
-        return bootstrap.are_resolved
+        bootstrap.resolve_until_success(now)
```
now=now right?

whirm replied:
oops, yes, thanks

NielsZeilemaker commented on tests/test_destroycommunity.py:
```diff
@@ -25,9 +32,11 @@ def test_hard_kill_without_permission(self):
+        message = node.create_full_sync_text("Should not be removed", 42)
         node.give_message(message, node)
 
+        node.process_packets()
```
shouldn't be neccesary

whirm replied:
let's try :D

NielsZeilemaker commented on callback.py:
```diff
@@ -803,3 +814,250 @@ def loop(self):
+                id_ = self.generate_id(id_)
+
+                # TODO(emilon): once register() is gone, we could use this instead of all this mess:
+                # result = blockingCallFromThread(reactor, call, *args, **kargs)
```
we can't really as blockingcall doesnt support a timeout

whirm replied:
hm, right

Commit:
Forgot to lower the log level for this

Commit:
convert ip to str

Commit:
force_dispersy_thread decorator

Commit:
attempt to fix unregister

Commit:
add the batch cache timeout callback to the community's pending callback list.

Commit:
Temporary hack until we fix the shutdown routine

Commit:
This should fix the race condition with the batch timeout callback

Commit:
twistify the batch cache

Commit:
This shouldn't be necessary

Commit:
Properly order the _register() args

Commit:
Add support for canceling all sorts of twisted primitives when unloading
the community.

Commit:
cleanup

Commit:
Twistify the endpoint

Commit:
Twistify test_overlay.py

Commit:
Twistify tracker.py

Commit:
Twistify main.py

LipuFei commented:
@whirm Could you do a rebase on devel? Then you'll be able to see clearly what I've done when I make a pull request to your branch. Thanks.

whirm commented:
@LipuFei Sure, il'l later

Commit:
twistify _download_master_member_identity

Commit:
Do not allow for callback ID's in pending callbacks anymore.

Commit:
Update comment

Commit:
make bootstrap.py callback free

Commit:
Twistify flush_database

Commit:
Twistify _stats_detailed_candidates

Commit:
The candidate walker stuf was moved to Community a while ago

Commit:
Twistify _periodically_clean_delayed

Commit:
Twistify take_step()

Commit:
Use a delayed call instead of the callback for the timeout event in the Cache

Commit:
Remove force_dispersy_thread decorator as it is not used

whirm commented:
@LipuFei I rebased a while ago. Right now I'm forking off devel's HEAD

Commit:
Twistify the periodic memory dumping feature.

Commit:
This statement wasn't true anymore

Commit:
request cache doesn't use the callback anymore

Commit:
community.py doesn't use the callback anymore

Commit:
dispersytestclass.py doesn't use the callback anymore

Commit:
node.py doesn't use the callback anymore

Commit:
Update doctrings

Commit:
Add a clear() method to the request cache so we can leave a completely
clean reactor when we stop a dispersy instance.

Commit:
fix/simplify Community._download_master_member_identity() to not leave
pending tasks behind.

Commit:
Revert/adapt "fix/simplify Community._download_master_member..." to a
static interval LoopingCall.

This reverts commit 0aaad467a7ef2f7cf2e4de0bef040462a9819ac2.

Commit:
Test that the reactor is clean before and after each individual test execution.

Commit:
Fully migrate dispersy.py away from the callback.

Commit:
Migrate main.py to twisted

Commit:
Port tracker.py to Twisted

Commit:
Remove callback.py and its tests, not used anymore.

NielsZeilemaker commented on community.py:
```diff
@@ -378,9 +388,12 @@ def on_dispersy_identity(message):
 
-        delay = 2.0
-        while not self._master_member.public_key:
+        def fetch_master_member_identity(delay):
```
Can we convert this one to a looping call?

NielsZeilemaker commented on dispersy.py:
```diff
@@ -169,17 +165,11 @@ def __init__(self, callback, endpoint, working_directory, database_filename=u"di
-                    scanner.dump_all_objects("memory-%d-shutdown.out" % (time() - start))
-
-            self._callback.register(memory_dump)
+            start = time()
```
This "logic" should be moved to the decorator class, and then lets add the @start_memory_dump at the constructor of dispersy

NielsZeilemaker replied:
Its still here, can you move it to the utils thingy

NielsZeilemaker commented on dispersy.py:
```diff
@@ -2316,135 +2284,134 @@ def start(self, timeout=10.0):
 
-        assert not self._callback.is_running, "Must be called before callback.start()"
+        # Disabled this assert as is_running now checks if the reactor is running
+        # assert not self._callback.is_running, "Must be called before callback.start()"
```
These two lines can be removed

NielsZeilemaker commented on dispersy.py:
```diff
@@ -2316,135 +2284,134 @@ def start(self, timeout=10.0):
-        results.append((u"callback", self._callback.start()))
         assert all(isinstance(result, bool) for _, result in results), [type(result) for _, result in results]
-        self._callback.call(start, priority=512)
+        assert isInIOThread(), "Must be called from the callback thread"
```
duplicate

NielsZeilemaker commented on endpoint.py:
```diff
@@ -354,7 +359,11 @@ def _loop(self):
                 finally:
                     if packets:
-                        logger.debug('%d came in, %d bytes in total', len(packets), sum(len(packet) for _, packet in packets))
+                        if logger and logger.debug:
```
This doesn't make sense, i'd rather not have checks for missing globals.
Let's address this by converting all loggers to a self._logger instead

whirm replied:
we are going to address this on a separate pull request (logger.py removal)

NielsZeilemaker commented on requestcache.py:
```diff
@@ -175,14 +154,14 @@ def add(self, cache):
             logger.debug("add %s", cache)
             self._identifiers[identifier] = cache
-            cache.callback_identifier = self._callback.register(self._on_timeout, (cache,), delay=cache.timeout_delay)
+            cache.timeout_delayed_call = reactor.callLater(cache.timeout_delay, self._on_timeout, cache)
```
Maybe we should opt for a second dictionary instead, Dispersy doesnt really use property assignmet as you're doing here.

NielsZeilemaker commented on tests/debugcommunity/node.py:
```diff
@@ -118,14 +126,18 @@ def init_my_member(self, tunnel=False, store_identity=True):
+            # TODO(emilon): Make it so it is no longer necessary to do this
             # remove introduction responses from socket
-            self.receive_messages(names=[u'dispersy-introduction-response'])
+            for _ in xrange(5):
```
Can we move this into the receive_messages?

NielsZeilemaker commented on tests/test_batch.py:
```diff
@@ -69,6 +69,7 @@ def test_one_big_batch(self, length=1000):
 
     def test_many_small_batches(self, length=1000):
         """
+        Test many small batches of messages is processed correctly.
```
are

NielsZeilemaker commented on tests/test_bootstrap.py:
```diff
@@ -14,13 +15,15 @@
         Runs tracker.py and connects to it.
         """
+
+        Bootstrap.enable = True
```
Is it set to false anywhere?

whirm replied:
on the test class module

NielsZeilemaker commented on tests/test_dynamicsettings.py:
```diff
@@ -46,6 +47,8 @@ def test_change_resolution(self):
         node.give_message(message, self._mm)
         other.give_message(message, self._mm)
 
+        #other.process_packets()
```
Remove this line

NielsZeilemaker commented on tests/test_overlay.py:
```diff
@@ -60,6 +68,40 @@ def take_step(self):
             def dispersy_enable_fast_candidate_walker(self):
                 return enable_fast_walker
 
+            @inlineCallbacks
```
Merge bug, the dispersy_enable_fast_candidate_walker property enabled the fast walker

NielsZeilemaker replied:
This still needs to be removed

Commit:
_pending_tasks is a dict now, implemented cancel_pending_task(),
_batch_cache doesn't store the Cache DelayedCall anymore as it was
stored in _pending_tasks too.

Commit:
Typo and cleanup

Commit:
Use a dict for the request timeout delayed calls

Commit:
Fix the call on reactor thread decorator

Commit:
Move the retrial loop to receive_messages so all the tests can benefit
from it.

Commit:
Fix and move the reactor call decorator to util.py and refactor all the
unit tests to use it.

Commit:
Convert test_nat_detection.py to unix newlines

NielsZeilemaker commented on bootstrap.py:
```diff
@@ -125,60 +127,44 @@ def reset(self):
 
-    def resolve(self, func, timeout=60.0, blocking=False):
+    @inlineCallbacks
+    def resolve(self, timeout=60.0):
```
The timeout argument isn't used, i guess we should remove it

NielsZeilemaker commented on community.py:
```diff
@@ -28,19 +34,19 @@
-
 logger = get_logger(__name__)
 
+DOWNLOAD_MM_PK_INTERVAL = 300.0
```
Can you reduce the interval to 15.0

NielsZeilemaker commented on community.py:
```diff
@@ -187,17 +194,17 @@ def __init__(self, dispersy, master, my_member, attach=True):
-        # community is unloaded.  most of the time this contains all the generators that are being
-        # used by the community
-        self._pending_callbacks = []
+        # _pending_callbacks contains all pending calls that should be removed when the
```
update to pending_tasks

NielsZeilemaker commented on community.py:
```diff
@@ -378,27 +388,23 @@ def on_dispersy_identity(message):
                 assert message.authentication.member.mid == self._master_member.mid
                 self._master_member = message.authentication.member
                 assert self._master_member.public_key
+                self._pending_tasks.pop("download master member identity").stop()
```
maybe call cancel_pending_task instead

NielsZeilemaker commented on community.py:
```diff
@@ -378,27 +388,23 @@ def on_dispersy_identity(message):
+                logger.debug("%s found master member", self._cid.encode("HEX"))
+                self._master_member = self._dispersy.get_member(public_key=str(public_key))
+                assert self._master_member.public_key
+                self._pending_tasks.pop("download master member identity").stop()
```
same here

NielsZeilemaker commented on dispersy.py:
```diff
@@ -68,12 +73,11 @@
-from .endpoint import Endpoint
+FLUSH_DATABASE_INTERVAL = 60.0
+STATS_DETAILED_CANDIDATES_INTERVAL = 5.0
+MEMORY_DUMP_INTERVAL = float(60 * 60)
```
This one should be moved to utils

NielsZeilemaker commented on dispersy.py:
```diff
@@ -300,36 +291,23 @@ def on_results(success):
-                # unable to resolve all... retry until successful
-                self._callback.register(retry_until_success, delay=300.0)
+            # first attempt will block for at most TIMEOUT seconds
+            result = yield bootstrap.resolve(timeout=timeout).addCallback(on_results)
```
Resolve doesn't support a timeout anymore, which is unfortunate

NielsZeilemaker commented on tests/debugcommunity/node.py:
```diff
@@ -189,6 +190,7 @@ def send_message(self, message, candidate):
 
         return self.send_packet(message.packet, candidate)
 
+    #@call_on_dispersy_thread
```
remove this line

NielsZeilemaker commented on util.py:
```diff
@@ -0,0 +1,55 @@
+# util.py ---
```
I know you like these headers, but lets remove it for now.
If we're going to add a license, all file need a similar header, or we can simply add a license file.
- can you put all other decorators in here as well?

Commit:
Cleanup

Commit:
Set the master member download interval to 15 secs

Commit:
Move all the decorators and the memory dumper to the util.py module.

whirm commented:
I'm still running the tests locally, but everything looks fine so far.

NielsZeilemaker commented on tests/test_overlay.py:
```diff
@@ -56,9 +65,39 @@ def take_step(self):
-            @property
-            def dispersy_enable_fast_candidate_walker(self):
-                return enable_fast_walker
+            @inlineCallbacks
```
This one should be removed, and the property reinstated

Commit:
Use the Community fast walker functionality

Commit:
Remove all remains of bootstrap timeout functionality.

Commit:
Port the inactive community cleanup loop to twisted
